gcsfs/core.py

```diff
@@ -931,17 +963,29 @@ def _fetch(self, start, end):
             self.cache = _fetch_range(self.gcsfs.header, self.details,
                                       start, self.end)
         if start < self.start:
-            new = _fetch_range(self.gcsfs.header, self.details,
-                               start, self.start)
-            self.start = start
-            self.cache = new + self.cache
+            if self.end - end > self.blocksize:
+                self.start = start
+                self.end = end + self.blocksize
+                self.cache = _fetch_range(self.gcsfs.header, self.details,
+                                          self.start, self.end)
+            else:
+                new = _fetch_range(self.gcsfs.header, self.details,
+                                   start, self.start)
+                self.start = start
+                self.cache = new + self.cache
         if end > self.end:
             if self.end > self.size:
                 return
-            new = _fetch_range(self.gcsfs.header, self.details,
-                               self.end, end + self.blocksize)
-            self.end = end + self.blocksize
-            self.cache = self.cache + new
+            if end - self.end > self.blocksize:
+                self.start = start
+                self.end = end + self.blocksize
+                self.cache = _fetch_range(self.gcsfs.header, self.details,
+                                          self.start, self.end)
+            else:
+                new = _fetch_range(self.gcsfs.header, self.details,
+                                   self.end, end + self.blocksize)
+                self.end = end + self.blocksize
+                self.cache = self.cache + new
```

Comment:
Why do we always add self.blocksize to end.  What if `end - start > self.blocksize`?

Reply:
The idea is to read past the currently requested data by a blocksize, so that further (small) reads will not need additional requests.

Reply:
We might do something like `max(start + self.blocksize, end)` so that we don't do this behavior on large reads.

Reply:
Sure, could do, however the little extra may make little difference for longer reads

Reply:
Fair enough